cmd/webhook-server/main.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright (c) 2019 StackRox Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"crypto/tls"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	tlsDir      = `/etc/webhook`
+	tlsCertFile = `tls.crt`
+	tlsKeyFile  = `tls.key`
+)
+
+var (
+	podResource         = metav1.GroupVersionResource{Version: "v1", Resource: "pods"}
+	cephclusterResource = metav1.GroupVersionResource{Group: "ceph.rook.io", Version: "v1", Resource: "cephclusters"}
+	dataDirHostPath string
+)
+
+func validateResource(req *v1beta1.AdmissionRequest) ([]patchOperation, error) {
+
+	var patches []patchOperation
+
+	log.Println(req.Resource)
+
+	if req.Resource == cephclusterResource && req.Operation == "CREATE" {
+		raw := req.Object.Raw
+		cephcluster := cephv1.CephCluster{}
+
+		if _, _, err := universalDeserializer.Decode(raw, nil, &cephcluster); err != nil {
+			return nil, fmt.Errorf("could not deserialize pod object: %v", err)
+		}
+
+		dataDirHostPath = cephcluster.Spec.DataDirHostPath
+
+	}
+
+	if req.Resource == cephclusterResource && req.Operation == "UPDATE" {
+		raw := req.Object.Raw
+		cephcluster := cephv1.CephCluster{}
+
+		if _, _, err := universalDeserializer.Decode(raw, nil, &cephcluster); err != nil {
+			return nil, fmt.Errorf("could not deserialize pod object: %v", err)
+		}
+
+		if cephcluster.Spec.DataDirHostPath != dataDirHostPath {
+			fmt.Println(cephcluster.ObjectMeta.Name + "\n" + cephcluster.ObjectMeta.Namespace)
+			return nil, errors.New("dataDirHostPath change after creation is not allowed !")
+		}
+	}
+
+	return patches, nil
+}
+
+func main() {
+
+
+	cert := flag.String("tls-cert-file", "cert.pem", "File containing the default x509 Certificate for HTTPS.")
+	key := flag.String("tls-private-key-file", "key.pem", "File containing the default x509 private key matching --tls-cert-file.")
+
+	flag.Parse()
+
+
+
+	keyPair, err := NewTlsKeypairReloader(*cert, *key)
+
+	if err != nil {
+		log.Printf("error load certificate: %v", err.Error())
```

Comment:
After logging the error, we have to return (or exit with an error code) from the main function

Reply:
Added to exit with an error code